File: src/AAS_Cores/Numerical_AAS_Core/src/main/java/utilities/InteractionsUtils.java
package utilities;

import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.*;
import java.util.Collections;
import java.util.Properties;

public class InteractionsUtils {


    // ------------------------
    // Methods related to files
    // ------------------------
    public static KafkaConsumer createInteractionKafkaConsumer() {

        Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KafkaInfo.KAFKA_SERVER_IP + ":9092");
        props.put(ConsumerConfig.CLIENT_ID_CONFIG, "i4-0-smia-core");
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());

        KafkaConsumer<String, String> kafkaConsumerPartitionManager = new KafkaConsumer<>(props);
        kafkaConsumerPartitionManager.assign(Collections.singletonList(
                new TopicPartition(KafkaInfo.KAFKA_TOPIC, KafkaInfo.MANAGER_TOPIC_PARTITION)));
        return kafkaConsumerPartitionManager;
    }

    public static String sendInteractionMsgToManager(String msg_key, JSONObject msg_data) {

        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, KafkaInfo.KAFKA_SERVER_IP + ":9092");
        props.put(ProducerConfig.CLIENT_ID_CONFIG, "i4-0-smia-core");
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());

        KafkaProducer<String, String> kafkaProducerPartitionCore = new KafkaProducer<>(props);
        ProducerRecord<String, String> record = new ProducerRecord<>(KafkaInfo.KAFKA_TOPIC,
                KafkaInfo.CORE_TOPIC_PARTITION, msg_key, msg_data.toString());
        kafkaProducerPartitionCore.send(record);
        kafkaProducerPartitionCore.flush();

        return "OK";
    }

    public static void changeStatus(String newStatus) {
        JSONObject statusFileJSON = fileToJSON(AAS_ArchiveInfo.coreStatusFilePath);
        statusFileJSON.put("status", newStatus);
        updateFile(AAS_ArchiveInfo.coreStatusFilePath, statusFileJSON);
    }

    public static String getManagerStatus() {
        JSONParser parser = new JSONParser();
        try (Reader reader = new FileReader(AAS_ArchiveInfo.managerStatusFilePath)) {
            JSONObject statusFileJSON = (JSONObject) parser.parse(reader);
            return statusFileJSON.get("status").toString();
        }  catch (FileNotFoundException e) {
            return null;
        } catch (IOException | ParseException e) {
            throw new RuntimeException(e);
        }
    }

    public static JSONObject fileToJSON(String filePath) {
        JSONParser parser = new JSONParser();
        try (Reader reader = new FileReader(filePath)) {
            return (JSONObject) parser.parse(reader);
        } catch (IOException | ParseException e) {
            throw new RuntimeException(e);
        }
    }

    public static void updateFile(String filePath, JSONObject content) {
        try (FileWriter file = new FileWriter(filePath)) {
            file.write(content.toJSONString());
            file.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    // -----------------------------------
    // Methods related to service requests
    // -----------------------------------
    public static JSONObject getNextSvcRequest() {
        /**
         * This method gets the next service request that has not be realized. To achieve that it analyzes the service
         * requests of AAS Manager, and check if there is an response for each request in the AAS Core, because if
         * there is not, it means that the request has to be performed. Returns the JSON object of the new request or
         * null if no request has been made.
         * @return JSON object with the information of the new service. NULL if it no request has been made.
         */

        // For each service, check if there is an response for each request, because if there is not, it means that the
        // request has to be performed
        JSONArray requestsArray = (JSONArray) fileToJSON(AAS_ArchiveInfo.managerInteractionsFolderPath + AAS_ArchiveInfo.svcRequestFileSubPath).get("serviceRequests");
        for (Object obj: requestsArray) {
            JSONObject reqObj = (JSONObject) obj;
            if (getResponseServiceJSON(((Long) reqObj.get("interactionID")).intValue()) == null)
                return reqObj;
        }
        return null;
    }

    // ------------------------------------
    // Methods related to service responses
    // ------------------------------------
    public static void setResponseServiceState(int interactionID, String newState) {
        // Conseguimos el JSON del servicio en el archivo de respuestas
        JSONObject serviceJSON = getResponseServiceJSON(interactionID);
        if (serviceJSON == null) {
            serviceJSON = new JSONObject();
            serviceJSON.put("interactionID", interactionID);
        }
        serviceJSON.put("serviceStatus", newState);
        updateSvcResponse(serviceJSON);
    }

    // ---------------------------------------
    // Methods related to responses (by AAS Core)
    //----------------------------------------
    private static void updateSvcResponse(JSONObject serviceJSON) {
        JSONArray responsesArray = (JSONArray) fileToJSON(AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath).get("serviceResponses");
        responsesArray.add(serviceJSON);
        JSONObject updatedContent = new JSONObject();
        updatedContent.put("serviceResponses", responsesArray);
        updateFile(AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath, updatedContent);
    }

    private static JSONObject getResponseServiceJSON(int interactionID) {
        JSONArray responsesArray = (JSONArray) fileToJSON(AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath).get("serviceResponses");
        for (Object o: responsesArray) {
            JSONObject svcJSON = (JSONObject) o;
            if (interactionID == ((Long) svcJSON.get("interactionID")).intValue())
                return svcJSON;
        }
        return null;
    }

    public static JSONObject createSvcCompletedResponse(JSONObject requestJSON, String serviceData) {
        JSONObject completedResponseJSON = new JSONObject();
        completedResponseJSON.put("interactionID", requestJSON.get("interactionID"));
        completedResponseJSON.put("serviceID", requestJSON.get("serviceID"));
        completedResponseJSON.put("serviceType", requestJSON.get("serviceType"));
        completedResponseJSON.put("serviceStatus", "Completed");
        if (serviceData != null) {
            JSONObject serviceDataJSON = new JSONObject();
            String requestedData = (String) ((JSONObject) requestJSON.get("serviceData")).get("requestedData");
            serviceDataJSON.put(requestedData, serviceData);
            serviceDataJSON.put("timestamp", System.currentTimeMillis() / 1000);   // Add the timestamp in seconds
            completedResponseJSON.put("serviceData", serviceDataJSON);
        }
        return completedResponseJSON;
    }

    public static void updateSvcCompleteResponse(JSONObject responseFinalJSON) {
        String svcResponsesFilePath = AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath;
        System.out.println(fileToJSON(svcResponsesFilePath).toJSONString());
        JSONArray responsesArray = (JSONArray) fileToJSON(svcResponsesFilePath).get("serviceResponses");
        if (responsesArray == null)
            responsesArray = new JSONArray();
        responsesArray.add(responseFinalJSON);
        JSONObject updatedResponseContent = new JSONObject();
        updatedResponseContent.put("serviceResponses", responsesArray);
        updateFile(svcResponsesFilePath, updatedResponseContent);
    }
}

File: use_cases/AAS_Cores/physical_asset_cores/OPC_UA_AAS_Core/src/utilities/Interactions_utils.py
"""
File to group useful methods related to interactions between AAS Manager and AAS Core.
"""
import calendar
import json
import time

from kafka import KafkaConsumer, TopicPartition, KafkaProducer
from kafka.errors import KafkaError

from utilities.AASArchive_utils import file_to_json, update_json_file
from utilities.AASarchiveInfo import AASarchiveInfo
from utilities.KafkaInfo import KafkaInfo


def get_next_svc_request():
    print("Obtaining the next service request from AAS Manager.")
    svc_requests_file_path = AASarchiveInfo.MANAGER_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_REQUEST_FILE_SUBPATH
    svc_requests_json = file_to_json(svc_requests_file_path)
    if len(svc_requests_json['serviceRequests']) != 0:
        return svc_requests_json['serviceRequests'][len(svc_requests_json['serviceRequests']) - 1]
    else:
        return None


def delete_svc_request(svc_request_object):
    print("Deleting the service request with ID " + str(svc_request_object['interactionID']) + " from AAS Archive.")
    svc_requests_file_path = AASarchiveInfo.MANAGER_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_REQUEST_FILE_SUBPATH
    svc_requests_json = file_to_json(svc_requests_file_path)
    for i in range(len(svc_requests_json['serviceRequests'])):
        if svc_requests_json['serviceRequests'][i]["interactionID"] == svc_request_object['interactionID']:
            svc_requests_json['serviceRequests'].pop(i)
            break
    update_json_file(svc_requests_file_path, svc_requests_json)

def create_response_json_object(svc_request_info, svc_params = None):

    response_json =  {
        'interactionID': svc_request_info['interactionID'],
        'thread': svc_request_info['thread'],
        'serviceType': svc_request_info['serviceType'],
        'serviceID': svc_request_info['serviceID'],
        'serviceData': {
            'serviceStatus': 'Completed',
            'serviceCategory': 'service-response',
            'timestamp': calendar.timegm(time.gmtime())
        }
    }
    if svc_params is not None:
        response_json['serviceData']['serviceParams'] = svc_params
    return response_json

def add_new_svc_response(new_response_json):
    """
    This method adds a new service response to the related service interaction file of the core and updates it
    in the AAS Archive.

    Args:
        new_response_json: The service requests content in JSON format.
    """

    # Get the content of the service requests interaction file
    svc_requests_file_path = AASarchiveInfo.CORE_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_RESPONSE_FILE_SUBPATH
    svc_requests_json = file_to_json(svc_requests_file_path)
    if svc_requests_json is None:
        svc_requests_json = {'serviceResponses': [new_response_json]}
    else:
        svc_requests_json['serviceResponses'].append(new_response_json)

    update_json_file(svc_requests_file_path, svc_requests_json)

# ---------------------
# Kafka related methods (Intra AAS interaction platform)
# ---------------------
def send_interaction_msg_to_manager(client_id, msg_key, msg_data):
    """
    This method sends a Kafka interaction message to the AAS Manager. To this end, the AAS Core will publish messages
    in its partition, where the AAS Core will be listening.

    Args:
        client_id (str): the id of the client of the Kafka producer.
        msg_key (str): the key of the Kafka message.
        msg_data: the data of the Kafka message.

    Returns:
        str: shipment status
    """
    # First, the Kafka producer is created
    kafka_producer = KafkaProducer(bootstrap_servers=[KafkaInfo.KAFKA_SERVER_IP + ':9092'],
                                   client_id=client_id,
                                   value_serializer=lambda x: json.dumps(x).encode('utf-8'),
                                   key_serializer=str.encode,
                                   )
    print("Kafka producer created. Sending message to AAS Manager through topic " + KafkaInfo.KAFKA_TOPIC +
          " and partition " + str(KafkaInfo.CORE_TOPIC_PARTITION) + "...")
    result = kafka_producer.send(KafkaInfo.KAFKA_TOPIC, value=msg_data, key=msg_key,
                                 partition=KafkaInfo.CORE_TOPIC_PARTITION)
    print("Message sent to topic " + KafkaInfo.KAFKA_TOPIC)
    print(str(result))
    try:
        record_metadata = result.get(timeout=10)
        print("Kafka producer has sent the message. Producer closing...")
        kafka_producer.close()
        return "OK"
    except KafkaError as e:
        print("ERROR")
        print(e)
        kafka_producer.close()
        return None


def create_interaction_kafka_consumer(client_id):
    """
    This method creates the Kafka consumer for subscribing to AAS Manager partition in the topic of the AAS.
    Args:
        client_id (str): the id of the client of the Kafka consumer.

    Returns:
        KafkaConsumer: the object of the Kafka consumer.
    """

    kafka_consumer_manager_partition = KafkaConsumer(bootstrap_servers=[KafkaInfo.KAFKA_SERVER_IP + ':9092'],
                                                     client_id=client_id,
                                                     value_deserializer=lambda x: json.loads(x.decode('utf-8')),
                                                     enable_auto_commit=True,
                                                     group_id='i4-0-smia-cores'
                                                     )
    kafka_consumer_manager_partition.assign([TopicPartition(KafkaInfo.KAFKA_TOPIC, KafkaInfo.MANAGER_TOPIC_PARTITION)])

    return kafka_consumer_manager_partition
